pkg/messaging/msgtopic.go

```diff
@@ -0,0 +1,151 @@
+	if len(msgParts) < numGroups {
+		return "", "", "", ErrMalformedTopic
+	}
+
+	domainID := msgParts[domainGroup]
+	chanID := msgParts[channelGroup]
+	subtopic := msgParts[subtopicGroup]
+
+	var err error
+	subtopic, err = ParsePublishSubtopic(subtopic)
+	if err != nil {
+		return "", "", "", errors.Wrap(ErrMalformedTopic, err)
+	}
+
+	return domainID, chanID, subtopic, nil
+}
+
+func ParsePublishSubtopic(subtopic string) (string, error) {
+	if subtopic == "" {
+		return subtopic, nil
+	}
+
+	subtopic, err := url.QueryUnescape(subtopic)
+	if err != nil {
+		return "", errors.Wrap(ErrMalformedSubtopic, err)
+	}
+	subtopic = strings.ReplaceAll(subtopic, "/", ".")
+
+	elems := strings.Split(subtopic, ".")
+	filteredElems := []string{}
+	for _, elem := range elems {
+		if elem == "" {
+			continue
+		}
+
+		if strings.Contains(elem, " ") || strings.Contains(elem, "#") || strings.Contains(elem, "+") || strings.Contains(elem, "*") || strings.Contains(elem, ">") {
+			return "", ErrMalformedSubtopic
+		}
+
+		filteredElems = append(filteredElems, elem)
+	}
+
+	subtopic = strings.Join(filteredElems, ".")
+	return subtopic, nil
+}
+
+func ParseSubscribeTopic(topic string) (string, string, string, error) {
+	msgParts := msgTopicRegExp.FindStringSubmatch(topic)
+	if len(msgParts) < numGroups {
+		return "", "", "", ErrMalformedTopic
+	}
+
+	domainID := msgParts[domainGroup]
+	chanID := msgParts[channelGroup]
+	subtopic := msgParts[subtopicGroup]
+
+	var err error
+	subtopic, err = ParseSubscribeSubtopic(subtopic)
+	if err != nil {
+		return "", "", "", errors.Wrap(ErrMalformedTopic, err)
+	}
+
+	return domainID, chanID, subtopic, nil
+}
+
+func ParseSubscribeSubtopic(subtopic string) (string, error) {
+	if subtopic == "" {
+		return subtopic, nil
+	}
+
+	subtopic = strings.ReplaceAll(subtopic, "/", ".")
+	subtopic = strings.ReplaceAll(subtopic, "+", "*")
+	subtopic = strings.ReplaceAll(subtopic, "#", ">")
+	elems := strings.Split(subtopic, ".")
+	filteredElems := []string{}
+	for _, elem := range elems {
+		if elem == "" {
+			continue
+		}
+
+		if strings.Contains(elem, " ") || (len(elem) > 1 && (strings.Contains(elem, "#") || strings.Contains(elem, "+") || strings.Contains(elem, "*") || strings.Contains(elem, ">"))) {
+			return "", ErrMalformedSubtopic
+		}
+
+		filteredElems = append(filteredElems, elem)
+	}
+
+	subtopic = strings.Join(filteredElems, ".")
+	return subtopic, nil
+}
+
+func EncodeToInternalSubject(domainID string, channelID string, subtopic string) string {
+	return fmt.Sprintf("%s.%s", MsgTopicPrefix, EncodeToInternalSubjectSuffix(domainID, channelID, subtopic))
+}
+
+func EncodeToInternalSubjectSuffix(domainID string, channelID string, subtopic string) string {
+	subject := fmt.Sprintf("%s.%s.%s", domainID, ChannelTopicPrefix, channelID)
+	if subtopic != "" {
+		subject = fmt.Sprintf("%s.%s", subject, subtopic)
+	}
+	return subject
+}
+
+func (m *Message) EncodeToInternalSubject() string {
```

Comment:
Let's not attach any new methods to the generated code - use `EncodeToInternalTopic` directly.

Reply:
Let's not use `internal` notion - `EncodeTopic`.